File: python/src/gwp_py/session.py
"""GQL session management."""

from __future__ import annotations

from typing import Any

import grpc

from gwp_py._convert import value_to_proto
from gwp_py._generated import gql_service_pb2 as gql_pb2
from gwp_py._generated import gql_service_pb2_grpc as gql_grpc
from gwp_py._generated import gql_types_pb2 as types_pb2
from gwp_py.errors import SessionError
from gwp_py.result import ResultCursor
from gwp_py.transaction import Transaction


class GqlSession:
    """An active session with a GWP server."""

    def __init__(
        self,
        session_id: str,
        session_stub: gql_grpc.SessionServiceStub,
        gql_stub: gql_grpc.GqlServiceStub,
    ):
        self._session_id = session_id
        self._session_stub = session_stub
        self._gql_stub = gql_stub
        self._closed = False

    @classmethod
    async def create(cls, channel: grpc.aio.Channel) -> GqlSession:
        """Create a new session via handshake."""
        session_stub = gql_grpc.SessionServiceStub(channel)
        gql_stub = gql_grpc.GqlServiceStub(channel)

        resp = await session_stub.Handshake(
            gql_pb2.HandshakeRequest(protocol_version=1)
        )

        if not resp.session_id:
            raise SessionError("server returned empty session ID")

        return cls(resp.session_id, session_stub, gql_stub)

    @property
    def session_id(self) -> str:
        """The session identifier."""
        return self._session_id

    async def execute(
        self,
        statement: str,
        parameters: dict[str, Any] | None = None,
    ) -> ResultCursor:
        """Execute a GQL statement."""
        proto_params = {}
        if parameters:
            for k, v in parameters.items():
                proto_params[k] = value_to_proto(v, types_pb2)

        stream = self._gql_stub.Execute(
            gql_pb2.ExecuteRequest(
                session_id=self._session_id,
                statement=statement,
                parameters=proto_params,
            )
        )

        return ResultCursor(stream)

    async def begin_transaction(self, *, read_only: bool = False) -> Transaction:
        """Begin a new transaction."""
        mode = gql_pb2.READ_ONLY if read_only else gql_pb2.READ_WRITE
        return await Transaction.begin(self._session_id, self._gql_stub, mode)

    async def set_graph(self, name: str) -> None:
        """Set the current graph."""
        await self._session_stub.Configure(
            gql_pb2.ConfigureRequest(session_id=self._session_id, graph=name)
        )

    async def set_schema(self, name: str) -> None:
        """Set the current schema."""
        await self._session_stub.Configure(
            gql_pb2.ConfigureRequest(session_id=self._session_id, schema=name)
        )

    async def set_time_zone(self, offset_minutes: int) -> None:
        """Set the session timezone."""
        await self._session_stub.Configure(
            gql_pb2.ConfigureRequest(
                session_id=self._session_id,
                time_zone_offset_minutes=offset_minutes,
            )
        )

    async def reset(self) -> None:
        """Reset session state to defaults."""
        await self._session_stub.Reset(
            gql_pb2.ResetRequest(session_id=self._session_id, target=gql_pb2.RESET_ALL)
        )

    async def ping(self) -> int:
        """Ping the server. Returns a timestamp."""
        resp = await self._session_stub.Ping(
            gql_pb2.PingRequest(session_id=self._session_id)
        )
        return resp.timestamp

    async def close(self) -> None:
        """Close the session."""
        if not self._closed:
            await self._session_stub.Close(
                gql_pb2.CloseRequest(session_id=self._session_id)
            )
            self._closed = True

    async def __aenter__(self) -> GqlSession:
        return self

    async def __aexit__(self, *args: object) -> None:
        await self.close()

File: src/client/session.rs
//! Client-side session wrapper.

use std::collections::HashMap;

use tonic::transport::Channel;

use crate::error::GqlError;
use crate::proto;
use crate::proto::gql_service_client::GqlServiceClient;
use crate::proto::session_service_client::SessionServiceClient;
use crate::types::Value;

use super::result::ResultCursor;
use super::transaction::Transaction;

/// An active session with a GQL server.
///
/// Wraps the handshake response and provides typed methods for
/// executing statements, managing transactions, and configuring
/// session state.
pub struct GqlSession {
    session_id: String,
    session_client: SessionServiceClient<Channel>,
    gql_client: GqlServiceClient<Channel>,
}

impl GqlSession {
    /// Create a new session by performing a handshake.
    pub(crate) async fn new(channel: Channel) -> Result<Self, GqlError> {
        let mut session_client = SessionServiceClient::new(channel.clone());
        let gql_client = GqlServiceClient::new(channel);

        let resp = session_client
            .handshake(proto::HandshakeRequest {
                protocol_version: 1,
                credentials: None,
                client_info: HashMap::new(),
            })
            .await?
            .into_inner();

        Ok(Self {
            session_id: resp.session_id,
            session_client,
            gql_client,
        })
    }

    /// Get the session ID.
    #[must_use]
    pub fn session_id(&self) -> &str {
        &self.session_id
    }

    /// Execute a GQL statement and return a cursor over the results.
    ///
    /// # Errors
    ///
    /// Returns an error if the server rejects the request.
    pub async fn execute(
        &mut self,
        statement: &str,
        parameters: HashMap<String, Value>,
    ) -> Result<ResultCursor, GqlError> {
        let proto_params: HashMap<String, proto::Value> = parameters
            .into_iter()
            .map(|(k, v)| (k, proto::Value::from(v)))
            .collect();

        let stream = self
            .gql_client
            .execute(proto::ExecuteRequest {
                session_id: self.session_id.clone(),
                statement: statement.to_owned(),
                parameters: proto_params,
                transaction_id: None,
            })
            .await?
            .into_inner();

        Ok(ResultCursor::new(stream))
    }

    /// Begin an explicit transaction.
    ///
    /// # Errors
    ///
    /// Returns an error if the transaction cannot be started.
    pub async fn begin_transaction(&mut self) -> Result<Transaction, GqlError> {
        Transaction::begin(
            self.session_id.clone(),
            self.gql_client.clone(),
            proto::TransactionMode::ReadWrite,
        )
        .await
    }

    /// Begin a read-only transaction.
    ///
    /// # Errors
    ///
    /// Returns an error if the transaction cannot be started.
    pub async fn begin_read_only_transaction(&mut self) -> Result<Transaction, GqlError> {
        Transaction::begin(
            self.session_id.clone(),
            self.gql_client.clone(),
            proto::TransactionMode::ReadOnly,
        )
        .await
    }

    /// Set the current graph for this session.
    ///
    /// # Errors
    ///
    /// Returns an error if the server rejects the configuration.
    pub async fn set_graph(&mut self, graph: &str) -> Result<(), GqlError> {
        self.session_client
            .configure(proto::ConfigureRequest {
                session_id: self.session_id.clone(),
                property: Some(proto::configure_request::Property::Graph(graph.to_owned())),
            })
            .await?;
        Ok(())
    }

    /// Set the current schema for this session.
    ///
    /// # Errors
    ///
    /// Returns an error if the server rejects the configuration.
    pub async fn set_schema(&mut self, schema: &str) -> Result<(), GqlError> {
        self.session_client
            .configure(proto::ConfigureRequest {
                session_id: self.session_id.clone(),
                property: Some(proto::configure_request::Property::Schema(
                    schema.to_owned(),
                )),
            })
            .await?;
        Ok(())
    }

    /// Set the timezone offset for this session.
    ///
    /// # Errors
    ///
    /// Returns an error if the server rejects the configuration.
    pub async fn set_time_zone(&mut self, offset_minutes: i32) -> Result<(), GqlError> {
        self.session_client
            .configure(proto::ConfigureRequest {
                session_id: self.session_id.clone(),
                property: Some(proto::configure_request::Property::TimeZoneOffsetMinutes(
                    offset_minutes,
                )),
            })
            .await?;
        Ok(())
    }

    /// Reset all session state to defaults.
    ///
    /// # Errors
    ///
    /// Returns an error if the server rejects the request.
    pub async fn reset(&mut self) -> Result<(), GqlError> {
        self.session_client
            .reset(proto::ResetRequest {
                session_id: self.session_id.clone(),
                target: proto::ResetTarget::ResetAll.into(),
            })
            .await?;
        Ok(())
    }

    /// Ping the server to check connectivity.
    ///
    /// # Errors
    ///
    /// Returns an error if the server is unreachable.
    pub async fn ping(&mut self) -> Result<i64, GqlError> {
        let resp = self
            .session_client
            .ping(proto::PingRequest {
                session_id: self.session_id.clone(),
            })
            .await?
            .into_inner();

        Ok(resp.timestamp)
    }

    /// Close this session.
    ///
    /// # Errors
    ///
    /// Returns an error if the server rejects the request.
    pub async fn close(mut self) -> Result<(), GqlError> {
        self.session_client
            .close(proto::CloseRequest {
                session_id: self.session_id.clone(),
            })
            .await?;
        Ok(())
    }
}
